### DISCUSSION TITLE
能否提供基于事件、消息驱动的二次开发方式

### DISCUSSION THREAD

**Discussion by dev4u** [shadowsocks/shadowsocks-rust#430] [Ideas]
我想对sslocal、ssserver做二次开发，以实施在我脑海里的构想。
**消息驱动：**
例如我想实现sslocal在建立连接后，从ssserver读取tcp、udp超时时长，发送、接收缓存大小……到本地，并以此初始化后续的连接。
**事件驱动：**
例如我想知道在ssserver何时收到重放数据？这样能定义对应的响应行为：回应随机数据亦或不做响应……或者可以唤醒定义的脚本。

不知这样是否可行？

**Comment by DuckSoft** [comment]
"您好,我有一个很棒的想法,只欠程序员"
"您好,我就是程序员,可惜我也有很棒的想法"

**Comment by dev4u** [comment]
不知sslocal、ssserver能不能改造出二次开发的接口来用呢？

**Reply by zonyitoo** [reply]
是的你可以这样做，用shadowsocks-service，或者直接用shadowsocks这个crate

**Reply by dev4u** [reply]
> 是的你可以这样做，用shadowsocks-service，或者直接用shadowsocks这个crate

你说的没毛病，但是直接修改这两个crate，不能算提供二次开发的方式。既没有回调，也没有注入的能力。

**Reply by zonyitoo** [reply]
不是修改，而是你用你的程序来引用shadowsocks这个crate，这个crate只是shadowsocks的协议

**Reply by dev4u** [reply]
> 不是修改，而是你用你的程序来引用shadowsocks这个crate，这个crate只是shadowsocks的协议

这样做，相当于又造了一个轮子出来，新轮和原轮只是气门芯的制式不一样而已。

我意思是这个你做的轮子，对气门部件做改造，使得既能用你官方的气门，也能用规格相同的其他气门。

**Reply by zonyitoo** [reply]
那你想要怎么样的设计呢

**Reply by dev4u** [reply]
或者我具体化问题：我想实现在ss rust中增加一种加密、解密的算法，我要怎么最小化“侵入”你的代码来实现？

**Reply by zonyitoo** [reply]
Hmm, there should be some kind of `Adapter` design.

**Reply by dev4u** [reply]
是的，我希望ss rust能实现adapter pattern和observer pattern。这样可以让大量的朋友加入进来，而不至于因为提的建议，不符合ss协议规范，只能干看着使不上劲。
